Firestore/Source/Remote/FSTStream.mm

```diff
@@ -394,8 +397,9 @@ - (void)closeWithFinalState:(FSTStreamState)finalState error:(nullable NSError *
     [self notifyStreamInterruptedWithError:error];
   }
 
-  // Clear the delegates to avoid any possible bleed through of events from GRPC.
-  _delegate = nil;
+  // PORTING NOTE: notifyStreamInterruptedWithError may have restarted the stream with a new
```

Comment:
This is the same problem we keep having: we can't notify our delegate in an inconsistent state. `[self notifyStreamInterruptedWIthError:error]` must be the last action we take here.

We've actually have observed that callback filter is ineffective. There's a race condition it doesn't handle, which is why https://github.com/firebase/firebase-ios-sdk/pull/790 was required.

The whole thing needs to be redesigned, but in the meantime I think we should clear the delegate, but just above the notification.

Reply:
Ah, wow.  Thanks for flagging this.  I didn't know there was history here.  I assume the problem is that the callbackFilter is applied /before/ the messages are dispatched to our queue and so even though we suppress callbacks in this code there could still be callbacks already waiting on our queue.

I don't think setting delegate to nil above notifyStreamInterruptedWithError helps any, since in the case I'm fixing the delegate is getting immediately re-assigned to a non-nil value (by notifyStreamInterruptedWithError). So those pending callbacks would just be delivered to the new delegate, which is exactly what we want to prevent with the callbackFilter. :-/

I think the fix is probably to make callbackFilter do the dispatch onto the dispatch queue in addition to doing the filtering so it can actually do its job.

Reply:
I guess the other options would be to remove the dispatch in `writeValue:` (there's a "TODO(mcg): remove the double-dispatch once GRPCCall at head is released.")...  If you think we can actually do that, that would also be sufficient to fix the race condition if I'm understanding it correctly.

Reply:
Sorry I had to implement it convince myself it would work, so it to you in a [PR](https://github.com/firebase/firebase-ios-sdk/pull/811).

1.  You're right, we can't nil out the delegate above the call to `notifyStreamInterruptedWithError:` because the delegate must be non-nil for that to even do anything useful. The two interesting implementations nil it out for themselves anyway.
2.  Getting gRPC to call us back on our dispatch queue fixes the race condition in `FSTCallbackFilter` so perhaps new further action is required.